### PR TITLE
[REVIEW] Optimizations of kernel launch configurations for apply functions

### PR DESCRIPTION
**Pull request by kkraus14**
In testing this gave a ~15x speedup for `apply_rows` with a sufficiently large DataFrame and basic UDF. For `apply_chunks` removes the guesswork in supplying `tpb` / `blkct` parameters and lets CUDA handle the kernel configuration by default (can still override if desired).

### CODE REVIEW

File: python/cudf/cudf/utils/applyutils.py [hunk 1]
```diff
@@ -148,7 +161,7 @@ def run(self, df, **launch_params):
         # Prepare output frame
         outdf = df.copy()
         for k in sorted(self.outcols):
-            outdf[k] = outputs[k]
+            outdf[k] = Series(outputs[k], nan_as_null=False)
```
Comment (beckernick) [inline python/cudf/cudf/utils/applyutils.py]: This is different nan/null behavior than before, AFAICT. Any specific reason?

File: python/cudf/cudf/utils/applyutils.py [hunk 2]
```diff
@@ -177,19 +188,24 @@ def compile(self, func, argnames, extra_argnames):
         )
         return kernel
 
-    def launch_kernel(self, df, args, chunks, tpb):
+    def launch_kernel(self, df, args, chunks, blkct=None, tpb=None):
```
Comment (beckernick) [inline python/cudf/cudf/utils/applyutils.py]: Would be nice to inject these new configuration into an existing apply-kernel test (or duplicate one)

File: python/cudf/cudf/utils/applyutils.py (hunk 1 above)
Comment (kkraus14) [reply python/cudf/cudf/utils/applyutils.py]: I figured if a user explicitly passes in data with `NaN` values from my perspective they wouldn't expect to get `null` values back. There's still the pessimistic null handling for handling actual nulls.

File: python/cudf/cudf/utils/applyutils.py (hunk 2 above)
Comment (kkraus14) [reply python/cudf/cudf/utils/applyutils.py]: Done

File: python/cudf/cudf/utils/applyutils.py (hunk 1 above)
Comment (beckernick) [reply python/cudf/cudf/utils/applyutils.py]: Good change.

File: python/cudf/cudf/tests/test_cuda_apply.py [hunk 3]
```diff
@@ -83,8 +83,7 @@ def test_df_apply_custom_chunks(nelem):
     def kernel(in1, in2, in3, out1, out2, extra1, extra2):
         for i, (x, y, z) in enumerate(zip(in1, in2, in3)):
             out1[i] = extra2 * x - extra1 * y + z
-            # cuda.blockDim.x is 1
-            out2[i] = i * cuda.blockDim.x
+            out2[i] = i
```
Comment (thomcom) [inline python/cudf/cudf/tests/test_cuda_apply.py]: Why drop `cuda.blockDim.x` here?

File: python/cudf/cudf/tests/test_cuda_apply.py [hunk 4]
```diff
@@ -118,8 +117,9 @@ def kernel(in1, in2, in3, out1, out2, extra1, extra2):
 
 
 @pytest.mark.parametrize("nelem", [1, 15, 30, 64, 128, 129])
-@pytest.mark.parametrize("tpb", [1, 8, 16, 64])
-def test_df_apply_custom_chunks_tpb(nelem, tpb):
+@pytest.mark.parametrize("blkct", [None, 1, 8])
+@pytest.mark.parametrize("tpb", [1, 8, 64])
+def test_df_apply_custom_chunks_blkct_tpb(nelem, blkct, tpb):
     def kernel(in1, in2, in3, out1, out2, extra1, extra2):
         for i in range(cuda.threadIdx.x, in1.size, cuda.blockDim.x):
```
Comment (thomcom) [inline python/cudf/cudf/tests/test_cuda_apply.py]: but not here?

File: python/cudf/cudf/tests/test_cuda_apply.py (hunk 3 above)
Comment (kkraus14) [reply python/cudf/cudf/tests/test_cuda_apply.py]: Because that was specifically written assuming the `tpb` setting was `1` which it no longer is.

**Pull request merged**